Add wallet classes

astrojarred opened the pull request:
Wanted to get the ball rolling on this! Lots of basic functionality is now here, along with the first few tests and examples. See /examples/wallet.py for some examples of the most basic funcionality.

Current capabilities:
- [x] Generate keys
- [x] Load keys
- [x] Query utxos
- [x] Send ada
- [x] Send specific UTxOs
- [x] Get senders of all UTxOs
- [x] Get metadata for all held tokens
- [x] Get utxo block times and sort utxos
- [x] Mint / Burn tokens
- [x] Automatically load in token polices where wallet is a signer
- [x] Automatically create BlockFrost Chain Context (mainnet, preprod, and preview)
- [x] Attach messages to transactions
- [x] Sign messages
- [x] Add custom metadata fields
- [x] Multi-output transactions
- [x] Register wallet
- [x] Stake wallet
- [x] Withdraw rewards
- [x] Generate fully manual transactions that can do any / all of the above

Future additions:
- [ ] Add more tests
- [ ] Add more examples
- [ ] Draft and sign multi-sig transactions
- [ ] Interaction with native scripts
- [ ] Interaction with plutus scripts
- [ ] Load wallets with mnemonic phrase

cffls commented on pycardano/wallet.py:
```diff
@@ -0,0 +1,1938 @@
+            return {}
+
+    else:
+        logger.warn(
+            "Retrieving stake address information is only possible with Blockfrost Chain Context."
+        )
```
Can we raise an exception when the wrong context is passed? Raising an explicit exception is probably safer than returning a None value. What do you think?

astrojarred replied:
Agreed. There are a few places where similar logic also might apply:

`Wallet.stake_info` and other stake-related methods warn when stake address is not yet registered. Not sure if it makes more sense to leave it as a warning and return nothing or raise and error of some sort.

`Wallet.sync()` warns when an address has not transacted on the chain yet. I think leaving it as a warning should be fine.
``

cffls replied:
Makes sense, for these two cases we can use warning. Thank you!

cffls reviewed (commented):
Thank you @astrojarred for this amazing work! This will make many things much easier to use for many people. The PR looks pretty good already. I added a few comments about the implementation and API naming. Please feel free to disagree. Again, thank you so much for your contribution! 🙏

astrojarred commented:
Hi Jerry,

Thanks for the helpful review. I concur with everything suggested and I especially like using the decorator for blockfrost-only contexts. I'll make the changes soon. 

And my apologies about the formatting, I often forget 😅

Thanks again!

astrojarred commented on pycardano/wallet.py:
```diff
@@ -0,0 +1,1938 @@
+    Returns:
+        dict: Info regarding the given stake address.
+    """
+    if isinstance(context, BlockFrostChainContext):
```
Nice. I will implement like so, such that context as both an arg and a kwarg are checked, as well as raising an error when a different context type is provided.

```python
def blockfrost_only(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        if isinstance(kwargs.get("context"), BlockFrostChainContext) or isinstance(args[1], BlockFrostChainContext):
            func(*args, **kwargs)
        else:
            raise TypeError(f"Function {func.__name__} is only available for context of type BlockFrostChainContext.")

    return wrapper
```

astrojarred reviewed (commented):
In reference to commit [c30fba1](https://github.com/cffls/pycardano/pull/89/commits/c30fba131e85a0eab96ffcadf7617142bbe5a6af):
@cffls I have the code generate wallet keys using the format:

```
name.skey
name.vkey
name.stake.skey
name.stake.vkey
```

This is to make it easy for a wallet to load unique staking keys. Let me know what you think of this format.

Thinking about this more I think I should also add inputs so one can directly specify the path to each for the four keyfiles individually.

cffls commented:
> In reference to commit [c30fba1](https://github.com/cffls/pycardano/pull/89/commits/c30fba131e85a0eab96ffcadf7617142bbe5a6af): @cffls I have the code generate wallet keys using the format:
> 
> ```
> name.skey
> name.vkey
> name.stake.skey
> name.stake.vkey
> ```
> 
> This is to make it easy for a wallet to load unique staking keys. Let me know what you think of this format.
> 
> Thinking about this more I now realize I should also add kwargs so one can directly specify the path to each for the four keyfiles individually.

The naming looks good to me. Thank you!

astrojarred commented:
Great, thanks! I believe I've addressed all the comments for now. Next step is to write more tests. I'll have to look in details at the ones you've already written for how to properly test functions that are meant to submit transactions to the chain.

cffls commented:
> Great, thanks! I believe I've addressed all the comments for now. Next step is to write more tests. I'll have to look in detail at the ones you've already written for how to properly test functions that are meant to submit transactions to the chain.

Nice! You can take a look at [integration tests](https://github.com/cffls/pycardano/tree/main/integration-test), where a local chain will be created through docker compose and you can run a bunch of real transactions on the chain.

daehan-koreapool commented:
FYI, the main branch now enforces stronger type hint in the Github ci/cd pipeline.

I would advise rebasing this branch on top of the latest main branch to ensure the newly added `wallet.py` module passes all static analyses.

astrojarred commented:
Done! I will work on straightening out the type hints, thank you!

robinboening commented:
Hey Jarred,

thanks for the big chunk of new features! Looks very promising to me!

However, I'd like to bring one thing up for a discussion...

You are introducing a new class `Wallet` as part of the public API. I like that because it is abstracting a lot of things which makes it easy to use.

Now here is what caught my eye. For some weeks now we also have a class `HDWallet`  https://github.com/Python-Cardano/pycardano/blob/88894e2cc97a8b01fed9a9fc99f2ff2861beac4b/pycardano/crypto/bip32.py#L78. Without looking into the details I would not understand why do we have `Wallet` and `HDWallet` and they don't provide the same / a similar interface.

First of all I am wondering what your thoughts are regarding the two in regards to the user/developer experience.

Secondly, I am wondering if it makes sense to extend the high-level `Wallet` class to support `HDWallet`.

Happy to hear what you're going to say :)

astrojarred commented on pycardano/wallet.py:
```diff
@@ -0,0 +1,1948 @@
+        return self.policy.policy_id
+
+    @blockfrost_only
+    def get_on_chain_metadata(self, context: ChainContext) -> dict:
```
Hi Daehan, 

Thanks for the input! As far as I know the methods marked with `@blockfrost_only` cannot be queried with Ogmios as they need information from db-sync. However, it is true that with the new typing system you've implemented the decorator is not needed anymore. It also means that I can simplify the type checking in many other places as well. So thank you for that!

astrojarred commented:
Hi @robinboening,

Thanks for pointing out the `HDWallet` class. I think it would be a really nice idea to integrate into `Wallet`. I think some might appreciate the option to create a wallet using a mnemonic or having full control over the generation of keys, so I don't think it's out of scope for this PR. I've updated the todo list above to specify the integration of `HDWallet`. In the end it might come in another PR later but I think it would be really nice to include.

astrojarred commented:
Quick update-- I've finally managed to integrate mypy so that the checks do not fail anymore. In addition I've gotten coverage up to ~44%. Still some tests to write but I'm closing in on this.
🔜

KINGH242 commented:
How can I help with this? I can't see the logs for the failed checks as it was too long ago. I am working on a project that would benefit from having the wallet stuff abstracted away into PyCardano.

cffls commented:
@KINGH242 that's great! This PR is staled and I completely forget about it.. I think a few type checks are failing. If those are fixed we should be able to merge it.

astrojarred commented:
Hey all, I dropped the ball a bit on this one!  

I think I got stuck with test writing-- I recall that I mainly had trouble with the mocking of the blockfrost API and tx submission as it's something I had no experience with.

Also happy to help out wherever possible and finally get this done.

KINGH242 commented:
Hey @astrojarred, I added some test and fixed a few things to allow it to work with the latest changes. I submitted a PR here https://github.com/astrojarred/pycardano/pull/33
